nucleus/annotation_uploader.py

```diff
@@ -0,0 +1,198 @@
+        local_files_per_upload_request: int = 10,
+        local_file_upload_concurrency: int = 30,
+    ):
+        if local_files_per_upload_request > 10:
+            raise ValueError("local_files_per_upload_request must be <= 10")
+        annotations_without_files: List[Annotation] = []
+        segmentations_with_local_files: List[SegmentationAnnotation] = []
+        segmentations_with_remote_files: List[SegmentationAnnotation] = []
+
+        for annotation in annotations:
+            if annotation.has_local_files():
+                # Only segmentations have local files currently, and probably for a long
+                # time to to come.
+                assert isinstance(annotation, SegmentationAnnotation)
+                segmentations_with_local_files.append(annotation)
+            elif isinstance(annotation, SegmentationAnnotation):
+                segmentations_with_remote_files.append(annotation)
+            else:
+                annotations_without_files.append(annotation)
+
+        responses = []
+        if segmentations_with_local_files:
+            responses.extend(
+                self.make_batched_file_form_data_requests(
+                    segmentations=segmentations_with_local_files,
+                    update=update,
+                    local_files_per_upload_request=local_files_per_upload_request,
+                    local_file_upload_concurrency=local_file_upload_concurrency,
+                )
+            )
+        if segmentations_with_remote_files:
+            # Segmentations require an upload and must be batched differently since a single
+            # segmentation will take a lot longer for the server to process than a single
+            # annotation of any other kind.
+            responses.extend(
+                self.make_batched_annotate_requests(
+                    segmentations_with_remote_files,
+                    update,
+                    batch_size=remote_files_per_upload_request,
+                    segmentation=True,
+                )
+            )
+        if annotations_without_files:
+            responses.extend(
+                self.make_batched_annotate_requests(
+                    annotations_without_files,
+                    update,
+                    batch_size=batch_size,
+                    segmentation=False,
+                )
+            )
+
+        return accumulate_dict_values(responses)
+
+    def make_batched_annotate_requests(
+        self,
+        annotations: Sequence[Annotation],
+        update: bool,
+        batch_size: int,
+        segmentation: bool,
+    ):
+        batches = [
+            annotations[i : i + batch_size]
+            for i in range(0, len(annotations), batch_size)
+        ]
+        responses = []
+        progress_bar_name = (
+            "Segmentation batches" if segmentation else "Annotation batches"
+        )
+        for batch in self._client.tqdm_bar(batches, desc=progress_bar_name):
```

Comment:
I think we should change the `tqdm_bar` to increment in `batch_sizes` instead of just counting batches, makes for a much nicer user experience.

Reply:
I can put that on the backlog but we follow this pattern in a lot of places right now, and this PR is already huge as is.

Reply:
https://app.shortcut.com/scaleai/story/407536/tqdm-progress-bars-should-show-item-level-not-batch-level-progress

Made a ticket for this